docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+monitoring configuration to be discovered and picked up with the use of the
+[Custom Resources][prom-crds] provided by the prometheus-operator. These
+resources can also be directly referenced in Go and be deployed by their
+respective components, instead of creating `.yaml` files in Go, or templating
+charts. With the addition of the prometheus-operator it should also be easier to
+add new features, such as Thanos.
+
+## Motivation
+
+Simplify monitoring changes and extensions with the use of the
+[prometheus-operator]. The current extension contract is described
+[here][extension-contract]. This document aims to define a new contract.
+
+Make it easier to add new monitoring features and make new changes. For example,
+when using the [prometheus-operator] components can bring their own monitoring
+configuration and specify exactly how they should be monitored without needing
+to add this configuration directly into Prometheus.
+
+The prometheus-operator handles validation of monitoring configuration. It will
+be more difficult to give prometheus invalid config.
+
+### Goals
+
+- Migrate from the current monitoring stack to the prometheus-operator.
+
+- Improve the monitoring extensibility and improve developer experience when
+  adding or editing configuration. This includes the monitoring extensions in
+  addition to core Gardener components.
+
+- Provide a clear direction on how monitoring resources should be managed.
+  Currently, there are many ways that monitoring configuration is being deployed
+  and this should be unified.
+
+- Leverage community knowledge by using [kube-prometheus] and
+  [kubernetes-mixins]. Reuse existing dashboards and monitoring configuration
+  for Kubernetes components whenever possible.
+
+- Improve how dashboards are discovered and provisioned for Grafana. Currently,
+  all dashboards are appended into a single configmap. This can be an issue if
+  ever exceed the maximum configmap size of 1MiB.
+
+### Non-Goals
+
+- Changes to the logging stack.
+
+- Feature parity between the current solution and the one proposed in this GEP.
+  The new stack should provide similar monitoring coverage, but it will be very
+  difficult to evaluate if there is feature parity. Perhaps some features will
+  be missing, but others may be added.
+
+## Proposal
+
+Today, Gardener provides monitoring for shoot clusters (i.e. system components
+and the control plane) and for the seed cluster. The proposal is to migrate
+these monitoring stacks to use the [prometheus-operator]. The proposal is lined
+out in the following steps:
+
+0. Define how will Gardener generate the Prometheus objects.
+    1. Option 1: Use the [API] provided by the [prometheus-operator] and create Go
+       structs.
+    2. Option 2: Template `yaml` files. The `yaml` should not be inline in Go,
+       but rather exist on its own. For some of the monitoring configuration it
+       is nice to look at the `yaml` representation rather than the Go structs.
```

Comment:
I think it would be good to decide on one option in this proposal and discard the other one.
Spoiler: the consensus will definitely be option 1 :)